bestla/bestla/bestla_device.h

```diff
@@ -303,10 +303,20 @@ class CpuDevice {
           for (auto& i : SMT_core) printf("%d,", i);
           printf("\n");
         }
-        E_L1Cache = L1[E_core[0]];
-        E_L2Cache = L2[E_core[0]] / 4;
-        L1Cache = E_L1Cache > L1[P_core[0]] / 2 ? L1[P_core[0]] / 2 : E_L1Cache;
-        L2Cache = E_L2Cache > L2[P_core[0]] / 2 ? L2[P_core[0]] / 2 : E_L2Cache;
+        if (!E_core.empty() && !P_core.empty()) {
+          E_L1Cache = L1[E_core[0]];
+          E_L2Cache = L2[E_core[0]] / 4;
+          L1Cache = E_L1Cache > L1[P_core[0]] / 2 ? L1[P_core[0]] / 2 : E_L1Cache;
+          L2Cache = E_L2Cache > L2[P_core[0]] / 2 ? L2[P_core[0]] / 2 : E_L2Cache;
+        } else if (!P_core.empty()) {
+          L1Cache = L1[P_core[0]] / 2;
+          L2Cache = L2[P_core[0]] / 2;
```

Comment:
p_core only == SMT ON?

Reply:
The P-core, E-core and SMT-core here only rely on CPU and P-core number always equal to SMT core number. We could not sure if it will use SMT when user bind cores. Using all cores include SMT will get the best performance on MTL. So we suppose user use SMT core by default now. It will be optimized after realizing threads pool and bonding core in threads pool (out of user's control).

Reply:
same question here, `!P_core.empty()` does not mean that SMT is used.

Reply:
This code can only detect whether SMT on in BIOS. And it cannot detect whether SMT used in our program. We don't even know how many thread user use to run our program here. P core list and SMT core list are used for core order in thread pool later.

Reply:
> This code can only detect whether SMT on in BIOS. And it cannot detect whether SMT used in our program. We don't even know how many thread user use to run our program here. P core list and SMT core list are used for core order in thread pool later.

then how line 287 works?

Reply:
line 287 is used to classify P-core/SMT and E-core/LPE-core. `core_id` record all the logic cores and its corresponding physical cores of CPU, regardless of whether they are used or not. P-core and SMT on the same physical core share one core id and four E-cores share one core id. This class describes physical CPU in design while `CPUBase` describes runtime CPU. `CPUBase` should handle all the things related to the thread number set by user.